### PR TITLE
Allow reduced max simultaneous TCP/UDP limits

### PR DESCRIPTION
**Pull request by jrowberg**
The ATWINC module firmware expects UDP sockets to begin at index 7 (right after the 0-6 allocated for TCP sockets). This means that if you want to reduce the number of simultaneous socket supported in your application firmware, e.g. for memory usage purposes, UDP sockets must still start at index 7 even if you support less than 7 TCP sockets.

To achieve this, I created an array of SOCKET handles (signed 8-bit integers) that maps sockets to buffer indexes. The socket array map always contains 11 elements (TCP_SOCK_MAX + UDP_SOCK_MAX). TCP sockets always match their indexes exactly, since they begin at 0. UDP sockets are mapped downward based on however many TCP sockets are omitted from the application.

In other words, normally UDP #0 will be index 7, but if you support exactly 2 TCP sockets instead of 7, then UDP #0 will be index 2.

Also included in this PR is the ability to allocate socket buffers at compile time, if desired. Deterministic memory allocation, especially when known at compile time, is valuable for mission-critical designs. This modification allows you to allocation socket buffers as arrays instead of using pointers and `malloc()`/`free()`.

Without any modifications to the new definitions near the top of WiFiSocket.h, the library retains its default behavior (dynamic memory allocation, 7 TCP sockets, 4 UDP sockets).

This PR addresses #242.

### CODE REVIEW

**Comment by sandeepmistry** [comment]
@jrowberg thanks for submitting this!

I'm thinking we hold off on these changes until there is more demand from the user base/community for static allocation and reduced sockets. @facchinm and @cmaglie what do you think?

**Comment by facchinm** [comment]
@sandeepmistry I'd give it a spin in mainline, with USE_STATIC_ALLOCATION undefined by default. With default defines it should have exactly the same behaviour of current version (except for the `_handleMap` indirection)

**Comment by sandeepmistry** [comment]
One more thought, if `TCP_SOCK_MAX` is reduced from 7, could this have a negative impact on server sockets, as the WINC1500 will decide what socket to use for the accepted connection.

**Comment by jrowberg** [comment]
@sandeepmistry That's a good question. If the module starts at 0 and uses the next available slot for additional TCP sockets (either incoming or outgoing), then it *should* be safe. However, I did not test this specifically. I'll try to do so and report back, though it might be a little while.

Thanks for considering it, regardless. I realize this is probably an uncommon use case.

**Comment by sandeepmistry** [comment]
I'm closing this for now due to lack of discussion, please re-open and comment if you are still interested.
